src/codeql.ts

```diff
@@ -0,0 +1,231 @@
+   */
+  databaseAnalyze(database: string, sarifFile: string, querySuite: string): Promise<void>;
+}
+
+export interface ResolveQueriesOutput {
+  byLanguage: {
+    [language: string]: {
+      [queryPath: string]: {}
+    }
+  };
+  noDeclaredLanguage: {
+    [queryPath: string]: {}
+  };
+  multipleDeclaredLanguages: {
+    [queryPath: string]: {}
+  };
+}
+
+/**
+ * Environment variable used to store the location of the CodeQL CLI executable.
+ * Value is set by setupCodeQL and read by getCodeQL.
+ */
+const CODEQL_ACTION_CMD = "CODEQL_ACTION_CMD";
+
+export async function setupCodeQL(): Promise<CodeQL> {
+  try {
+    const codeqlURL = core.getInput('tools', { required: true });
+    const codeqlURLVersion = getCodeQLURLVersion(codeqlURL);
+
+    let codeqlFolder = toolcache.find('CodeQL', codeqlURLVersion);
+    if (codeqlFolder) {
+      core.debug(`CodeQL found in cache ${codeqlFolder}`);
+    } else {
+      const codeqlPath = await toolcache.downloadTool(codeqlURL);
+      const codeqlExtracted = await toolcache.extractTar(codeqlPath);
+      codeqlFolder = await toolcache.cacheDir(codeqlExtracted, 'CodeQL', codeqlURLVersion);
+    }
+
+    let codeqlCmd = path.join(codeqlFolder, 'codeql', 'codeql');
+    if (process.platform === 'win32') {
+      codeqlCmd += ".exe";
+    } else if (process.platform !== 'linux' && process.platform !== 'darwin') {
+      throw new Error("Unsupported plaform: " + process.platform);
+    }
+
+    core.exportVariable(CODEQL_ACTION_CMD, codeqlCmd);
+    return getCodeQLForCmd(codeqlCmd);
+
+  } catch (e) {
+    core.error(e);
+    throw new Error("Unable to download and extract CodeQL CLI");
+  }
+}
+
+export function getCodeQLURLVersion(url: string): string {
+
+  const match = url.match(/\/codeql-bundle-(.*)\//);
+  if (match === null || match.length < 2) {
+    throw new Error(`Malformed tools url: ${url}. Version could not be inferred`);
+  }
+
+  let version = match[1];
+
+  if (!semver.valid(version)) {
+    core.debug(`Bundle version ${version} is not in SemVer format. Will treat it as pre-release 0.0.0-${version}.`);
+    version = '0.0.0-' + version;
+  }
+
+  const s = semver.clean(version);
+  if (!s) {
+    throw new Error(`Malformed tools url ${url}. Version should be in SemVer format but have ${version} instead`);
+  }
+
+  return s;
+}
+
+export function getCodeQL(): CodeQL {
+  const codeqlCmd = util.getRequiredEnvParam(CODEQL_ACTION_CMD);
+  return getCodeQLForCmd(codeqlCmd);
+}
+
+function getCodeQLForCmd(cmd: string): CodeQL {
+  return {
+    getDir: function() {
+      return path.dirname(cmd);
+    },
+    printVersion: async function() {
+      await exec.exec(cmd, [
+        'version',
+        '--format=json'
+      ]);
+    },
+    getTracerEnv: async function(database: string, compilerSpec: string | undefined) {
+      let envFile = path.resolve(database, 'working', 'env.tmp');
+      const compilerSpecArg = compilerSpec ? "--compiler-spec=" + compilerSpec : [];
```

Comment:
```suggestion
      const compilerSpecArg = compilerSpec ? ["--compiler-spec=" + compilerSpec] : [];
```

without the square brackets here, `...compilerSpecArg` a few lines later will split the string ""--compiler-spec=____" into an array of individual characters, which I don't think is what you want.

The deleted version of this function in `src/setup-tracer.ts` had the square brackets.

Reply:
Thank you. Good spot.